### PR TITLE
support new config & Feat config listener

### PR DESCRIPTION
**Pull request by chuntaojun**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/boot/discovery_watch.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import (
+	"context"
+	"time"
+)
+
+import (
+	"github.com/arana-db/arana/pkg/config"
+)
+
+//WatchTenants
```
Comment (Mulavar) [inline pkg/boot/discovery_watch.go:30]: put a space between // and method comment

File: pkg/config/event.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	EventTypeShadowRule
+)
+
+type (
+	Event interface {
+		Type() EventType
+	}
+
+	//TenantsEvent
+	TenantsEvent struct {
```
Comment (Mulavar) [inline pkg/config/event.go:39]: as above

File: pkg/config/api.go [hunk 3]
```diff
@@ -53,54 +62,85 @@ const (
...
-	return storeOperate.Init(options)
-}
+	Options struct {
+		StoreName string                 `yaml:"name"`
+		RootPath  string                 `yaml:"root_path"`
+		Options   map[string]interface{} `yaml:"options"`
+	}
 
-// Register register store plugin
-func Register(s StoreOperate) {
```
Comment (Mulavar) [inline pkg/config/api.go:80]: why rename Register to Regis? I think maybe the Register is a better method name. Just discuss.

**Review by Mulavar — APPROVED** [review]

File: pkg/boot/discovery_test.go [hunk 4]
```diff
@@ -19,6 +19,8 @@ package boot
 
 import (
 	"context"
+	"github.com/arana-db/arana/pkg/constants"
+	"os"
```
Comment (jjeffcaii) [inline pkg/boot/discovery_test.go:23]: reformat

File: pkg/boot/discovery_watch.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	cancel := fp.tenantOp.Subscribe(ctx, func(e config.Event) {
+		ch <- e.(config.TenantsEvent)
+	})
+
+	return ch, wrapWatchCancel(cancel, func() {
+		close(ch)
+	}), nil
+}
+
+//WatchNodes
```
Comment (jjeffcaii) [inline pkg/boot/discovery_watch.go:43]: Same as above

File: pkg/boot/proto.go [hunk 6]
```diff
@@ -166,17 +172,30 @@ type ConfigUpdater interface {
 	RemoveTable(ctx context.Context, tenant, cluster, table string) error
 }
 
+//ConfigWatcher
```
Comment (jjeffcaii) [inline pkg/boot/proto.go:175]: Insert a space between `//` and method name.

File: pkg/config/api.go [hunk 7]
```diff
@@ -31,16 +33,23 @@ type (
...
-const (
-	DefaultConfigPath                   PathKey = "/arana-db/config"
-	DefaultConfigMetadataPath           PathKey = "/arana-db/config/metadata"
-	DefaultConfigDataListenersPath      PathKey = "/arana-db/config/data/listeners"
-	DefaultConfigDataFiltersPath        PathKey = "/arana-db/config/data/filters"
-	DefaultConfigDataSourceClustersPath PathKey = "/arana-db/config/data/dataSourceClusters"
-	DefaultConfigDataShardingRulePath   PathKey = "/arana-db/config/data/shardingRule"
-	DefaultConfigDataTenantsPath        PathKey = "/arana-db/config/data/tenants"
+var (
+	_rootPathTemp      = "/%s/%s/"
```
Comment (jjeffcaii) [inline pkg/config/api.go:37]: Should be `const`.

File: pkg/config/api.go [hunk 8]
```diff
@@ -53,54 +62,85 @@ const (
...
 	storeOperate StoreOperate
+
+	once sync.Once
 )
 
-func GetStoreOperate() (StoreOperate, error) {
-	if storeOperate != nil {
-		return storeOperate, nil
+// Regis register store plugin
+func Regis(s StoreOperate) {
```
Comment (jjeffcaii) [inline pkg/config/api.go:73]: Why not `Register`?

File: pkg/config/api.go [hunk 9]
```diff
@@ -53,54 +62,85 @@ const (
...
+		Subscribe(ctx context.Context, c callback) context.CancelFunc
 	}
 
-	slots[s.Name()] = s
-}
+	// Center
+	Center interface {
+		io.Closer
+
+		//Load 每次都是拉取全量的配置数据
```
Comment (jjeffcaii) [inline pkg/config/api.go:117]: translate && insert a space

File: pkg/config/compare.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+ *
+ */
+
+package config
+
+import (
+	"reflect"
+)
+
+func (u *User) Compare(o *User) bool {
```
Comment (jjeffcaii) [inline pkg/config/compare.go:25]: How about `Equals`? Normally, `Compare` should return `-1,0,1`, just personal suggestion.

File: pkg/config/default.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package config
+
+import (
+	"github.com/arana-db/arana/pkg/util/log"
+	"github.com/pkg/errors"
```
Comment (jjeffcaii) [inline pkg/config/default.go:23]: reformat

File: pkg/config/etcd/etcd.go [hunk 12]
```diff
@@ -26,65 +26,80 @@ import (
...
+
 	if err != nil {
 		log.Errorf("failed to initialize etcd client error: %s", err.Error())
 		return err
 	}
 
-	c.client = tmpClient
-	c.lock = &sync.RWMutex{}
+	c.client = rawClient
+	c.lock = sync.RWMutex{}
```
Comment (jjeffcaii) [inline pkg/config/etcd/etcd.go:77]: No need to initialize `Mutex` or `RwMutex`.

**Review by jjeffcaii — COMMENTED** [review]
pls fix

**Comment by chuntaojun** [comment]
上述问题这两天会修复处理下，大家看看逻辑上是否有什么问题

File: pkg/config/nacos/nacos.go [hunk 13]
```diff
@@ -48,37 +48,37 @@ const (
...
+
+var (
+	PluginName = "nacos"
+
+	ErrorPublishConfigFail = errors.New("save config into nacos fail")
 )
 
 func init() {
-	config.Register(&storeOperate{})
+	config.Regis(&storeOperate{})
```
Comment (dongzl) [inline pkg/config/nacos/nacos.go:62]: why fix this name?

File: pkg/config/model.go [hunk 14]
```diff
@@ -61,13 +60,24 @@ type (
...
+	DataSourceType string
+
 	Data struct {
-		Filters            []*Filter            `yaml:"filters" json:"filters,omitempty"`
-		Listeners          []*Listener          `validate:"required,dive" yaml:"listeners" json:"listeners"`
-		Tenants            []*Tenant            `validate:"required,dive" yaml:"tenants" json:"tenants"`
-		DataSourceClusters []*DataSourceCluster `validate:"required,dive" yaml:"clusters" json:"clusters"`
-		ShardingRule       *ShardingRule        `validate:"required,dive" yaml:"sharding_rule,omitempty" json:"sharding_rule,omitempty"`
-		ShadowRule         *ShadowRule          `yaml:"shadow_rule,omitempty" json:"shadow_rule,omitempty"`
+		Filters []*Filter `yaml:"filters" json:"filters,omitempty"`
```
Comment (dongzl) [inline pkg/config/model.go:79]: `Filter` is removed at `master`.

File: pkg/config/file/file.go [hunk 15]
```diff
@@ -40,20 +43,58 @@ import (
...
 
-var configFilenameList = []string{"config.yaml", "config.yml"}
+var (
+	PluginName         = "file"
+	configFilenameList = []string{"config.yaml", "config.yml"}
+)
 
 func init() {
-	config.Register(&storeOperate{})
+	config.Regis(&storeOperate{})
```
Comment (dongzl) [inline pkg/config/file/file.go:52]: suggest keep the `Register` name.

File: pkg/config/diff.go [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+ */
+
+package config
+
+import (
+	"reflect"
+)
+
+func (t Tenants) Diff(old Tenants) *TenantsEvent {
+	newN := make([]string, 0, 4)
```
Comment (dongzl) [inline pkg/config/diff.go:26]: Suggest use the whole name `newTenants` & `deleteTenants`;

or not the suffix, only `new` and `delete`.

File: pkg/config/diff.go [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+	}
+
+	return &TenantsEvent{
+		AddTenants:    newN,
+		DeleteTenants: deleteN,
+	}
+}
+
+func (n Nodes) Diff(old Nodes) *NodesEvent {
+	newN := make([]*Node, 0, 4)
```
Comment (dongzl) [inline pkg/config/diff.go:59]: The same to above.

File: pkg/config/diff.go [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+
+	return &NodesEvent{
+		AddNodes:    newN,
+		UpdateNodes: updateN,
+		DeleteNodes: deleteN,
+	}
+}
+
+func (u Users) Diff(old Users) *UsersEvent {
+	newN := make([]*User, 0, 4)
```
Comment (dongzl) [inline pkg/config/diff.go:90]: The same to above.

File: pkg/config/diff.go [hunk 19]
```diff
@@ -0,0 +1,307 @@
...
+		AddUsers:    newN,
+		UpdateUsers: updateN,
+		DeleteUsers: deleteN,
+	}
+}
+
+func (c Clusters) Diff(old Clusters) *ClustersEvent {
+	newC := make([]*DataSourceCluster, 0, 4)
+	updateC := make([]*ClusterEvent, 0, 4)
+	deleteC := make([]*DataSourceCluster, 0, 4)
```
Comment (dongzl) [inline pkg/config/diff.go:134]: The same to above.

File: pkg/config/diff.go [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+		GroupsEvent: Groups(d.Groups).Diff(old.Groups),
+	}
+
+	return ret
+}
+
+func (g Groups) Diff(old Groups) *GroupsEvent {
+	newG := make([]*Group, 0, 4)
+	updateG := make([]*Group, 0, 4)
+	deleteG := make([]*Group, 0, 4)
```
Comment (dongzl) [inline pkg/config/diff.go:189]: The same to above.

File: pkg/config/diff.go [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+	}
+}
+
+func (s *ShardingRule) Diff(old *ShardingRule) *ShardingRuleEvent {
+	newT := make([]*Table, 0, 4)
+	updateT := make([]*Table, 0, 4)
+	deleteT := make([]*Table, 0, 4)
+
+	newTmp := map[string]*Table{}
+	oldTmp := map[string]*Table{}
```
Comment (dongzl) [inline pkg/config/diff.go:233]: The same to above.

File: pkg/config/diff.go [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+	}
+}
+
+func (s *ShadowRule) Diff(old *ShadowRule) *ShadowRuleEvent {
+	newT := make([]*ShadowTable, 0, 4)
+	updateT := make([]*ShadowTable, 0, 4)
+	deleteT := make([]*ShadowTable, 0, 4)
+
+	newTmp := map[string]*ShadowTable{}
+	oldTmp := map[string]*ShadowTable{}
```
Comment (dongzl) [inline pkg/config/diff.go:274]: The same to above.

**Review by dongzl — CHANGES_REQUESTED** [review]

File: pkg/config/compare.go (hunk 10 above)
Comment (chuntaojun) [reply pkg/config/compare.go:25]: ok

**Review by jjeffcaii — APPROVED** [review]
LGTM

File: conf/bootstrap.yaml [hunk 23]
```diff
@@ -15,19 +15,29 @@
...
+
+#   name: nacos
+#   options:
+#     endpoints: "127.0.0.1:8848"
+#     namespace: arana
+#     group: arana
+#     contextPath: /nacos
+#     scheme: http
+#     username: nacos
+#     password: nacos
```
Comment (dongzl) [inline conf/bootstrap.yaml:43]: add blank line.

File: testdata/fake_bootstrap.yaml [hunk 24]
```diff
@@ -15,66 +15,28 @@
...
+
+  # name: nacos
+  # options:
+  #   endpoints: "localhost:8080"
+  #   namespace: arana
+  #   group: arana
+  #   contextPath: /nacos
+  #   scheme: http
+  #   username: nacos
+  #   password: nacos
```
Comment (dongzl) [inline testdata/fake_bootstrap.yaml:42]: add blank line.

**Review by dongzl — CHANGES_REQUESTED** [review]

**Review by dongzl — APPROVED** [review]

**Pull request merged**